Testing and Deployment

kalyakinaa opened the pull request.

kalyakinaa committed:
Task done

kalyakinaa committed:
Forgotten file

SergeyK2004 commented on cypress/e2e/constructor/constructor.cy.tsx:
```diff
@@ -0,0 +1,103 @@
+
+const TEST_URL = 'http://localhost:4000';
+const SELECTORS = {
+  MODAL: '[data-cy="modal"]',
```
Отлично:
Хорошо, что вынесли селекторы в константы

SergeyK2004 commented on cypress/e2e/constructor/constructor.cy.tsx:
```diff
@@ -0,0 +1,103 @@
+
+const cleanupTestEnvironment = () => {
+  cy.clearLocalStorage();
+  cy.clearCookies();
```
Отлично:
Правильно, что очищаете куки и хранилище от моковых данных в конце теста

SergeyK2004 commented on cypress/e2e/constructor/constructor.cy.tsx:
```diff
@@ -0,0 +1,103 @@
+  });
+
+  describe('Модальные окна', () => {
+    Cypress._.times(15, (index) => {
```
Надо исправить:
Не следует зацикливать одинаковые тесты. Достаточно проверить работу приложения с одним ингредиентом.

SergeyK2004 commented on cypress/e2e/constructor/constructor.cy.tsx:
```diff
@@ -0,0 +1,103 @@
+
+  describe('Добавление ингредиентов', () => {
+    it('должно добавлять булки и начинки в конструктор', () => {
+      cy.get(SELECTORS.INGREDIENT_TYPE('bun')).contains('Добавить').click();
```
**Надо исправить:**

Перед тем, как добавлять ингредиент, нужно проверить, что его нет в конструкторе. В противном случае вы не проверяете действие, а проверяете только наличие ингредиента в конструкторе, но добавился он из-за клика в кнопку "добавить" или уже был там неизвестно.
При проверке любого действия необходимо выполнить три шага:

1. Проверить, что элемент, который должен появиться в результате действия отсутствует.
2. Выполнить действие.
3. Проверить, что элемент появился в разметке.

Необходимо учесть это во всех проверках.

SergeyK2004 commented on cypress/e2e/constructor/constructor.cy.tsx:
```diff
@@ -0,0 +1,103 @@
+      describe(`Тестирование ингредиента ${ingredientId}`, () => {
+        it('должно открывать и закрывать модальное окно', () => {
+          cy.get(SELECTORS.INGREDIENT(ingredientId)).click();
+          cy.get(SELECTORS.MODAL).should('be.visible');
```
Надо исправить:
Для проверки работы модального окна необходимо кликнуть в какой-то конкретный ингредиент и проверить отображение именно его данных в модальном окне

SergeyK2004 commented on cypress/e2e/constructor/constructor.cy.tsx:
```diff
@@ -0,0 +1,103 @@
+
+      cy.get(SELECTORS.CONSTRUCTOR.BUN(1)).contains('Выберите булки').should('exist');
+      cy.get(SELECTORS.CONSTRUCTOR.BUN(2)).contains('Выберите булки').should('exist');
+      cy.get(SELECTORS.CONSTRUCTOR.INGREDIENTS).contains('Выберите начинку').should('exist');
```
Отлично:
Хорошо, что проверили очистку конструктора

SergeyK2004 commented on src/services/tests/ingredientsSlice.test.ts:
```diff
@@ -0,0 +1,74 @@
+    it('должен корректно обрабатывать успешную загрузку ингредиентов', () => {
+      const newState = ingredientsReducer(
+        initialState,
+        getIngredients.fulfilled(mockIngredients, 'requestId')
```
Отлично:
Хорошие тесты асинхронных экшенов

SergeyK2004 reviewed (changes requested):
Очень хорошая работа. 
Тесты написаны аккуратно без лишнего кода. Код приятно читать. Все тесты проходят. 

Однако есть пара замечаний к логике тестов на cypress. Посмотрите их в коде на Гитхабе.

kalyakinaa committed:
Cypress tests corrections

Изменения cypress:
- добавлена проверка на отсутствие ингредиента перед его добавлением в конструктор,
- добавлен тест на корректность отображения данных (названия) добавленных в конструктор ингредиентов,
- проверка модальных окон производится на одном ингредиенте (удален цикл),
- добавлена проверка содержимого модального окна (в связи с этим изменен файл src\components\ui\ingredient-details\ingredient-details.tsx),
- добавлена проверка пустоты конструктора при оформлении заказа.

В файлах директории "src\services\tests" изменено форматирование.

SergeyK2004 reviewed (approved):
Отлично.
Замечаний больше нет.
Ваша работа принята.
